File: app/src/main/java/com/zibete/proyecto1/Adapters/GroupsDiffCallback.java
package com.zibete.proyecto1.Adapters;

import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.DiffUtil;

import com.zibete.proyecto1.model.Groups;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class GroupsDiffCallback extends DiffUtil.Callback {

    private ArrayList<Groups> mNewList;
    private ArrayList<Groups> mOldList;


    public GroupsDiffCallback(List<Groups> newlist, List<Groups> oldlist) {
        this.mNewList = (ArrayList<Groups>) newlist;
        this.mOldList = (ArrayList<Groups>) oldlist;
    }

    @Override
    public int getOldListSize() {
        return mOldList.size();
    }

    @Override
    public int getNewListSize() {
        return mNewList.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        return mOldList.get(oldItemPosition).getUsers()==mNewList.get(newItemPosition).getUsers();
    }


    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        return mOldList.get(oldItemPosition).equals(mNewList.get(newItemPosition));
    }

    @Nullable
    @Override
    public Object getChangePayload(int oldItemPosition, int newItemPosition) {
        Groups newItem = mNewList.get(newItemPosition);
        Groups oldItem = mOldList.get(oldItemPosition);

        Bundle diff = new Bundle();

        if(!Objects.equals(newItem.getUsers(), oldItem.getUsers())){

            diff.putString("users", String.valueOf(newItem.getUsers()));

        }

        if (diff.size()==0){
            return null;
        }
        return diff;
    }
}

File: app/src/main/java/com/zibete/proyecto1/adapters/GroupsDiffCallback.kt
package com.zibete.proyecto1.adapters

import androidx.recyclerview.widget.DiffUtil
import com.zibete.proyecto1.core.constants.Constants.PAYLOAD_GROUPS_DATA
import com.zibete.proyecto1.core.constants.Constants.PAYLOAD_GROUPS_USERS
import com.zibete.proyecto1.domain.roomsv2.RoomV2DirectoryItem

object GroupsDiffCallback : DiffUtil.ItemCallback<RoomV2DirectoryItem>() {

    override fun areItemsTheSame(
        oldItem: RoomV2DirectoryItem,
        newItem: RoomV2DirectoryItem,
    ): Boolean = oldItem.roomId == newItem.roomId

    override fun areContentsTheSame(
        oldItem: RoomV2DirectoryItem,
        newItem: RoomV2DirectoryItem,
    ): Boolean = oldItem == newItem

    override fun getChangePayload(
        oldItem: RoomV2DirectoryItem,
        newItem: RoomV2DirectoryItem,
    ): Any? {
        val changed = mutableSetOf<String>()

        if (oldItem.memberCount != newItem.memberCount) changed += PAYLOAD_GROUPS_USERS
        if (oldItem.description != newItem.description) changed += PAYLOAD_GROUPS_DATA

        return changed.takeIf { it.isNotEmpty() }
    }
}
